go/vt/discovery/healthcheck.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	stream, errfunc, err := conn.StreamHealth(ctx)
+	if err != nil {
+		conn.Close()
+		return nil, nil, err
+	}
+	hcc.mu.Lock()
+	hcc.conn = conn
+	hcc.mu.Unlock()
+	return stream, errfunc, nil
+}
+
+// processResponse reads one health check response, and notifies HealthCheckStatsListener.
+func (hcc *healthCheckConn) processResponse(ctx context.Context, hc *HealthCheck, endPoint *pbt.EndPoint, stream <-chan *pbq.StreamHealthResponse, errfunc tabletconn.ErrFunc) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case shr, ok := <-stream:
+		if !ok {
+			return errfunc()
+		}
+		if shr.Target == nil || shr.RealtimeStats == nil || shr.RealtimeStats.HealthError != "" {
+			return fmt.Errorf("health stats is not valid: %v", shr)
+		}
+
+		if hcc.target == nil {
+			// The first time we see response for the endpoint.
+			hcc.mu.Lock()
+			hcc.target = shr.Target
+			hcc.stats = shr.RealtimeStats
+			hcc.mu.Unlock()
+			hc.mu.Lock()
+			key := hc.endPointToMapKey(endPoint)
+			hc.addrToConns[key] = hcc
+			hc.addEndPointToTargetProtected(hcc.target, endPoint)
+			hc.mu.Unlock()
+		} else if hcc.target.TabletType != shr.Target.TabletType {
+			hc.mu.Lock()
+			hc.deleteEndPointFromTargetProtected(hcc.target, endPoint)
+			hcc.mu.Lock()
+			hcc.target = shr.Target
+			hcc.stats = shr.RealtimeStats
+			hcc.mu.Unlock()
+			hc.addEndPointToTargetProtected(shr.Target, endPoint)
+			hc.mu.Unlock()
+		} else {
+			hcc.mu.Lock()
+			hcc.target = shr.Target
+			hcc.stats = shr.RealtimeStats
+			hcc.mu.Unlock()
+		}
+		// notify downstream for tablettype and realtimestats change
+		hc.listener.StatsUpdate(endPoint, hcc.cell, hcc.target, hcc.stats)
+		return nil
+	}
+}
+
+// AddEndPoint adds the endpoint, and starts health check.
+func (hc *HealthCheck) AddEndPoint(cell string, endPoint *pbt.EndPoint) {
+	go hc.checkConn(cell, endPoint)
+}
+
+// RemoveEndPoint removes the endpoint, and stops the health check.
+func (hc *HealthCheck) RemoveEndPoint(endPoint *pbt.EndPoint) {
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+
+	key := hc.endPointToMapKey(endPoint)
+	hcc, ok := hc.addrToConns[key]
+	if !ok {
+		return
+	}
+	hcc.cancelFunc()
+	delete(hc.addrToConns, key)
+	if hcc.target != nil {
+		hc.deleteEndPointFromTargetProtected(hcc.target, endPoint)
+	}
+}
+
+// GetEndPointsFromKeyspaceShard returns all endpoints for the given keyspace/shard.
+func (hc *HealthCheck) GetEndPointsFromKeyspaceShard(keyspace, shard string) []*pbt.EndPoint {
+	hc.mu.RLock()
+	defer hc.mu.RUnlock()
+	shardMap, ok := hc.targetToEPs[keyspace]
+	if !ok {
+		return nil
+	}
+	ttMap, ok := shardMap[shard]
+	if !ok {
+		return nil
+	}
+	res := make([]*pbt.EndPoint, 0, 1)
+	for _, epList := range ttMap {
+		res = append(res, epList...)
+	}
+	return res
+}
+
+// GetEndPointsFromTarget returns all endpoints for the given target.
+func (hc *HealthCheck) GetEndPointsFromTarget(target *pbq.Target) []*pbt.EndPoint {
```

Comment:
I'm not a big fan of passing in structure pointers that contain the args of a method, I'd rather pass all three items in here, to avoid an unnecessary memory allocation...